panda/src/navigation/navMeshQuery.cxx

```diff
@@ -0,0 +1,229 @@
+/**
+ * PANDA 3D SOFTWARE
+ * Copyright (c) Carnegie Mellon University.  All rights reserved.
+ *
+ * All use of this software is subject to the terms of the revised BSD
+ * license.  You should have received a copy of this license along
+ * with this source code in a file named "LICENSE."
+ *
+ * @file navMeshQuery.cxx
+ * @author ashwini
+ * @date 2020-060-21
+ */
+
+#include "navMeshQuery.h"
+#include "navMeshNode.h"
+#include "DetourNavMeshQuery.h"
+#include "lvecBase3.h"
+#include <iostream>
+
+NavMeshQuery::NavMeshQuery():
+  _nav_query(0) {
+  _nav_query = dtAllocNavMeshQuery();
+}
+
+/**
+ * This function forms and initialises query object. No need to call set_nav_query() 
+ * after calling this constructor.
+ */
+NavMeshQuery::NavMeshQuery(PT(NavMesh) nav_mesh):
+  _nav_query(0) {
+  
+  _nav_query = dtAllocNavMeshQuery();
+  set_nav_query(nav_mesh);
+
+}
+NavMeshQuery::NavMeshQuery(NodePath nav_mesh_node_path):
+  _nav_query(0) {
+  
+  _nav_query = dtAllocNavMeshQuery();
+  set_nav_query(nav_mesh_node_path);
+
+}
+
+
+NavMeshQuery::~NavMeshQuery() {
+  dtFreeNavMeshQuery(_nav_query);
+}
+
+/**
+ * This function initializes the query object.
+ * It must be the first function called after construction, before other functions are used
+ * and can be used multiple times.
+ */
+bool NavMeshQuery::set_nav_query(PT(NavMesh) nav_mesh) {
+  const int MAX_NODES = 2048;
+  dtStatus status = _nav_query->init(nav_mesh->get_nav_mesh(), MAX_NODES);
+
+  if (dtStatusFailed(status)) {
+    std::cout<<"\nCannot set nav query!\n";
+    return false;
+  }
+
+  return true;
+}
+
+bool NavMeshQuery::set_nav_query(NodePath nav_mesh_node_path) {
+  NavMeshNode *node = dynamic_cast<NavMeshNode*>(nav_mesh_node_path.node());
```

Comment:
Use Panda's own casting system, which does not rely on RTTI.  To use it: `DCAST(NavMeshNode, nav_mesh_node_path.node())`